compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp

```diff
@@ -431,31 +602,88 @@ static void createCudaReductionStrategyStagedThreadDistribution(
   // single block/workgroup. Keep everything fused.
-  auto [maybeLeadingHFused, gridReductionH, maybeTiledTrailingH] =
+  auto [maybeLeadingHBlock, gridFillH, gridReductionH,
+        maybeTiledTrailingHBlock] =
       createReductionStrategyBlockDistribution(
           b, maybeLeadingH, fillH, reductionH, maybeTrailingH, strategy);
-  maybeLeadingH = maybeLeadingHFused;
 
   // Step 3. Split the reduction and tile the pieces to ensure vector
   // load/stores and mapping to a single warp with shuffles.
+  // TODO: consider fusing gridFillH.
   createReductionStrategyStagedThreadDistribution(
-      b, gridReductionH, maybeLeadingH, maybeTiledTrailingH, strategy);
+      b, gridReductionH, maybeLeadingHBlock, maybeTiledTrailingHBlock,
+      strategy);
 
-  // Step 4. Bufferize and drop HAL decriptor from memref ops.
-  Value funcH = b.create<MatchOp>(variantH, func::FuncOp::getOperationName());
-  funcH = iree_compiler::buildVectorize(b, funcH);
-  variantH = iree_compiler::buildBufferize(b, variantH,
-                                           /*targetGpu=*/true);
+  // Step 4-6. Common trailing steps.
+  createCommonTrailingStrategy(b, variantH, strategy);
+}
 
-  // Step 5. Post-bufferization mapping to blocks and threads.
-  // Need to match again since bufferize invalidated all handles.
-  // TODO: assumes a single func::FuncOp to transform, may need hardening.
-  funcH = b.create<MatchOp>(variantH, func::FuncOp::getOperationName());
-  funcH = iree_compiler::buildMapToBlockAndThreads(
-      b, funcH, strategy.getNumThreadsInBlock());
+static std::tuple<Value, Value, Value>
+createSmallReductionStrategyThreadDistribution(
+    ImplicitLocOpBuilder &b, Value maybeLeadingH, Value fillH, Value reductionH,
+    Value maybeTrailingH, const GPUReductionStrategy &strategy) {
+  auto pdlOperation = pdl::OperationType::get(b.getContext());
+  auto fusionTargetSelector = b.create<TakeFirstOp>(
+      pdlOperation, pdlOperation, ArrayRef<Value>{maybeTrailingH, reductionH});
+  Value fusionTargetH = fusionTargetSelector.getFirst();
+  Value fusionGroupH = fusionTargetSelector.getRest();
+  ArrayRef<Attribute> allThreadsRef(strategy.allThreadAttrs);
+  iree_compiler::TileToForeachThreadAndFuseAndDistributeResult tileResult =
+      iree_compiler::buildTileFuseDistToForeachThreadWithNumThreads(
+          /*builder=*/b,
+          /*rootH=*/fusionTargetH,
+          /*opsToFuseH=*/fusionGroupH,
+          /*numThreads=*/
+          getAsOpFoldResult(b.getI64ArrayAttr(strategy.workgroupTileSizes)),
+          /*threadDimMapping=*/
+          b.getArrayAttr(
+              allThreadsRef.take_front(strategy.captures.reductionRank - 1)));
+  fillH = b.create<FuseIntoContainingOp>(fillH, tileResult.foreachThreadH);
+  if (strategy.captures.maybeLeadingRank > 0) {
+    maybeLeadingH = b.create<FuseIntoContainingOp>(maybeLeadingH,
+                                                   tileResult.foreachThreadH);
+  }
```

Comment:
I'm not a fan of this. This makes the strategy too dynamic and difficult to express in the transform IR without adding control flow there, which we want to avoid. I'd rather play with empty handles.

Reply:
ok, fair enough, I'll refactor calls to "TakeFirstOp" which for some reason I always have troubles with in my mental model ..

Reply:
Fusion is robust to empty handles, you don't actually need control flow nor its "TakeFirstOp" equivalent here. Just always fuse the leading handle.